graphics/scaler/downscaler.cpp

```diff
@@ -62,11 +63,82 @@ void DownscaleAllByHalfTemplate(const uint8 *srcPtr, uint32 srcPitch, uint8 *dst
+void DownscaleAllByHalf(const uint8 *srcPtr, uint32 srcPitch, uint8 *dstPtr, uint32 dstPitch, int width, int height, int gBitFormat) {
 	if (gBitFormat == 565)
 		DownscaleAllByHalfTemplate<Graphics::ColorMasks<565> >(srcPtr, srcPitch, dstPtr, dstPitch, width, height);
 	else
 		DownscaleAllByHalfTemplate<Graphics::ColorMasks<555> >(srcPtr, srcPitch, dstPtr, dstPitch, width, height);
 }
 
 #endif
+
+namespace {
+uint32 getBrightness(byte col, const byte *pal) {
+	return pal[3 * col] * pal[3 * col] + pal[3 * col + 1] * pal[3 * col + 1] + pal[3 * col + 2] * pal[3 * col + 2];
+}
+
+void DownscaleCLUT8ByHalf(const uint8 *srcPtr, uint32 srcPitch, uint8 *dstPtr, uint32 dstPitch, int maxX, int maxY, const byte *palette) {
+	const byte *src1 = srcPtr;
+	const byte *src2 = srcPtr + srcPitch;
+	byte *dst = dstPtr;
+	int32 srcSkip = 2 * srcPitch - maxX * 2;
+	int32 dstSkip = dstPitch - maxX;
+	for (int y = 0; y < maxY; y++) {
+		for (int x = 0; x < maxX; x++) {
+			// Choose the brightest pixel. Writing often is bright on
+			// a dark background, so this preserves text as much as we can.
+			byte colors[4] = {
+				*src1++,
+				*src1++,
+				*src2++,
+				*src2++
+			};
+
+			byte col = colors[0];
+			uint32 bri = getBrightness(col, palette);
+			for (uint i = 1; i < 4; i++) {
+				uint32 nbri = getBrightness(colors[i], palette);
+				if (nbri > bri) {
+					bri = nbri;
+					col = colors[i];
+				}
+			}
+
+			*dst++ = col;
+		}
+		src1 += srcSkip;
+		src2 += srcSkip;
+		dst += dstSkip;
+	}
+}
+}
+
+void DownscaleSurfaceByHalf(Surface *out, const Surface *in, const byte *palette) {
+	if (in->format.isCLUT8() && out->format.isCLUT8() && palette != nullptr) {
+		DownscaleCLUT8ByHalf((const byte *) in->getBasePtr(0, 0), in->pitch, (byte *) out->getBasePtr(0, 0), out->pitch,
+				     MIN<int>(out->w, in->w / 2), MIN<int>(out->h, in->h / 2), palette);
+		return;
+	}
+
+	if ((in->format == PixelFormat(2, 5, 6, 5, 0, 11, 5, 0, 0)
+	     && out->format == PixelFormat(2, 5, 6, 5, 0, 11, 5, 0, 0))
+	    || (in->format == PixelFormat(2, 5, 6, 5, 0, 0, 5, 11, 0)
+		&& out->format == PixelFormat(2, 5, 6, 5, 0, 0, 5, 11, 0))) {
+		DownscaleAllByHalf((const byte *) in->getBasePtr(0, 0), in->pitch, (byte *) out->getBasePtr(0, 0), out->pitch,
+				   MIN<int>(in->w, out->w * 2), MIN<int>(in->h, out->h * 2), 565);
+		return;
+	}
+
+	if ((in->format == PixelFormat(2, 5, 5, 5, 0, 10, 5, 0, 0)
+	     && out->format == PixelFormat(2, 5, 5, 5, 0, 10, 5, 0, 0))
+	    || (in->format == PixelFormat(2, 5, 5, 5, 0, 0, 5, 10, 0)
+		&& out->format == PixelFormat(2, 5, 5, 5, 0, 0, 5, 10, 0))) {
+		DownscaleAllByHalf((const byte *) in->getBasePtr(0, 0), in->pitch, (byte *) out->getBasePtr(0, 0), out->pitch,
+				   MIN<int>(in->w, out->w * 2), MIN<int>(in->h, out->h * 2), 555);
+		return;
+	}
+
+	error("Unsupported downscale format %s->%s", in->format.toString().c_str(), out->format.toString().c_str());
```

Comment:
Please add prefix `downscaleCLUT8ByHalf(): ` to the message

Reply:
Done